### DISCUSSION TITLE
Kawai K1/K4

### DISCUSSION THREAD

**Discussion by nekton2** [christofmuc/KnobKraft-orm#294] [General]
I would like to create a K4 adaptation based on the K1, even if that doesn't make much sense due to the separation of patch and effects.  How exactly should I understand the "offsets" parameter?  I assume that the individual blocks (singles, multis, drums and effects) can be separated in a complete dump?

**Comment by christofmuc** [comment]
Very good question! I had to remind myself a bit, the K1 implementation is a very special one where I tried to implement the synth capabilities in a data driven way instead of pure code.

The offset parameter seems to be only partially implemented - line 112 shows its usage. I read this basically as the base address of the bank. So the first bank starts at 0, and then the next bank at 32 given a bank size of 32 and so on. The other two values don't appear to be used (yet).

The way I work with different data types right now is to separate them out into "banks". the newer "bankDescriptors" function allows to specify a patch type per bank, which ideally should be used to keep effects in effects banks and patches in pach banks etc., though I don't think anybody has actually used this so far.

The K1 adaptation is older thank bankDescriptors, so it doesn't use it yet. The Novation summit for example uses the bankDescriptors, and it should also be documented in the Adaptation programming guide.

I am not sure how good the K1 is as a base for implementing your own adaptation, given it is special and probably overly complicated. The K4 seems to be pretty straightforward except for the different patch types: https://erichizdepski.files.wordpress.com/2021/01/kawai-k4-wave-list-midi-impl.pdf

Maybe a better start would be the Matrix 1000 adaptation: https://github.com/christofmuc/KnobKraft-orm/blob/master/adaptations/Matrix1000.py

Implement the detection and the single request/single dump transfer first, that should get you started and we can see how to add effects and multis later (most adaptations don't support multis anyway...)

**Reply by nekton2** [reply]
Thanks for the explanations! The K4's SysEx format is completely transparent to me, so "only" transferring it to an adaptation is a challenge.  The K4 also has an edit buffer, which does not exist in the K1 and therefore in the adaptation.

**Reply by christofmuc** [reply]
Understand. The sysex stuff can be tricky to read and understand, it certainly helps to have done a few adaptations before so I know what to look for. Still, there is a lot of trial and error involved.

Attached a first version for a Kawai K4 - best case this allows you to auto detect the synth and query a bank via the import from synth... dialog. Both single and multi are supported, in theory in should be able to read also internal and cartridge memory.

Let me know what comes out of it, and remember to show the MIDI log you get here.
[Kawai K4.zip](https://github.com/christofmuc/KnobKraft-orm/files/13824395/Kawai.K4.zip)

**Reply by nekton2** [reply]
Thanks! I'll give it a try.

**Reply by christofmuc** [reply]
I added the K4 to the 2.2.0 release. It's the same version as attached here, but might be easier for you to test!

**Reply by nekton2** [reply]
Sure? There is only KawaiK1.py in the adaptions folder.

**Reply by christofmuc** [reply]
I must have forgotten to add it to the installer. You can download the file here: 
https://raw.githubusercontent.com/christofmuc/KnobKraft-orm/master/adaptations/Kawai%20K4.py
and put it into the user adaptation folder. Then restart the Orm and it should offer the K4.

Will make sure to include it properly in the next release!

**Comment by christofmuc** [comment]
Closing as solved.
